nixos/modules/services/web-servers/agate.nix

```
@@ -0,0 +1,122 @@
+        type = types.package;
+        default = pkgs.agate;
+        defaultText = literalExpression "pkgs.agate";
+        description = "The package to use";
+      };
+
+      addresses = mkOption {
+        type = types.listOf types.str;
+        default = [ "0.0.0.0:1965" ];
+        description = ''
+          Addresses to listen on, IP:PORT, if you haven't disabled forwarding
+          only set IPv4.
+        '';
+      };
+
+      contentDir = mkOption {
+        default = "/var/lib/agate/content";
+        type = types.path;
+        description = "Root of the content directory.";
+      };
+
+      certificatesDir = mkOption {
+        default = [ "/var/lib/agate/certificates" ];
+        type = types.listOf types.path;
+        description = "Root of the certificate directory.";
+      };
+
+      hostnames = mkOption {
+        default = [ ];
+        type = types.listOf types.str;
+        description = ''
+          Domain name of this Gemini server, enables checking hostname and port
+          in requests. (multiple occurences means basic vhosts)
+        '';
+      };
+
+      language = mkOption {
+        default = null;
+        type = types.nullOr types.str;
+        description = "RFC 4646 Language code for text/gemini documents.";
+      };
+
+      onlyTls_1_3 = mkOption {
+        default = false;
+        type = types.bool;
+        description = "Only use TLSv1.3 (default also allows TLSv1.2).";
+      };
+
+      extraArgs = mkOption {
+        type = types.str;
+        default = "";
+        example = "--log-ip";
+        description = "Extra arguments to use running agate.";
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.agate =
+      let
+        convertListArgs = key: list: concatMapStringsSep " " (value: "${key} '${value}'") list;
+        addresses = convertListArgs "--addr" cfg.addresses;
+        hostnames = convertListArgs "--hostname" cfg.hostnames;
+      in
+      {
+        description = "Agate";
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        preStart = ''
+          if [ ! -d '${cfg.contentDir}' ]; then
+            mkdir -p '${cfg.contentDir}'
+            chown -R agate:agate '${cfg.contentDir}'
+          fi
+          if [ ! -d '${cfg.certificatesDir}' ]; then
+            mkdir -p '${cfg.certificatesDir}'
+            chown -R agate:agate '${cfg.certificatesDir}'
+          fi
+        '';
+        serviceConfig = {
+          Type = "simple";
+          PermissionsStartOnly = true;
+          ExecStart = "${cfg.package}/bin/agate ${concatStringsSep " " ([
+            "--content '${cfg.contentDir}'"
+            "--certs ${cfg.certificatesDir}"
+            addresses
+            (optionalString (cfg.hostnames != []) hostnames)
+            (optionalString (cfg.language != null) "--lang '${cfg.language}'")
+            (optionalString cfg.onlyTls_1_3 "--only-tls13")
+            cfg.extraCommandLine
+          ])}";
```

this is long and complex enough that would benefit from being put into a script in the service itself instead of serviceConfig.ExecStart, eg

```suggestion
        script = ''
          exec ${cfg.package}/bin/agate ${escapeShellArgs [
            "--content" cfg.contentDir
            # ... and so on
          ]}
        '';
```

---

I tried script but I have no idea why it is failing

```
        script =
          let
            prefixKeyList = key: list: concatMap (v: [ key v ]) list;
            addresses = prefixKeyList "--addr" cfg.addresses;
            hostnames = prefixKeyList "--hostname" cfg.hostnames;
          in
          ''
            exec ${cfg.package}/bin/agate ${
              escapeShellArgs (
                [
                  "--content" "${cfg.contentDir}"
                  "--certs" "${cfg.certificatesDir}"
                ] ++
                addresses ++
                (optionals (cfg.hostnames != []) hostnames) ++
                (optionals (cfg.language != null) [ "--lang" cfg.language ]) ++
                (optionals cfg.onlyTls_1_3 [ "--only-tls13" ]) ++
                (optionals (cfg.extraArgs != []) cfg.extraArgs)
              )
            }
          '';
```

```
cleanup
kill machine (pid 8)
geminiserver # qemu-system-x86_64: terminating on signal 15 from pid 6 (/nix/store/97w52ckcjnfiz89h3lh7zf1kysgfm2s8-python3-3.9.6/bin/python3)
(0.02 seconds)
Traceback (most recent call last):
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 1318, in <module>
    driver.run_tests()
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 1194, in run_tests
    self.test_script()
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 1190, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 1, in <module>
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 536, in wait_for_unit
    retry(check_active)
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 192, in retry
    if fn(False):
  File "/nix/store/4q2aprlydycjgx1x7aw77x6p7lm0raqf-nixos-test-driver/bin/.nixos-test-driver-wrapped", line 528, in check_active
    raise Exception(
Exception: unit "agate" is inactive and there are no pending jobs
```

---

oh, we may have been unclear in the initial comment. apologies.

the `script` attribute does not go into `serviceConfig`, putting it there will indeed not work. `script` is an attribute at the same level as `serviceConfig`, ie it should be like
```
systemd.services.agate = {
  description = "Agate";
  script = #...
```

putting the script there has the test succeed for us.

---

oooh I see, I was putting it in `serviceConfig`

It's fairly cleaned up at this point though, is there much benefit moving between `ExecStart` & `script` ?

I can even move the args to a var and have the exec line be `ExecStart="${cfg.package}/bin/agate ${args}";`

---

`script` is an actual shell script, which means we can use `escapeShellArgs` to format arguments for it. `ExecStart` is not and has different escaping rules; using `script` when lots of escaping is involved just feels cleaner and more straightforward.

---

Ah ok, I'll sort that now